packages/auth-sveltekit/src/server.ts

```diff
@@ -0,0 +1,489 @@
+import {
+  redirect,
+  type Cookies,
+  type RequestEvent,
+  type Handle,
+} from "@sveltejs/kit";
+import type { Client } from "edgedb";
+import {
+  Auth,
+  builtinOAuthProviderNames,
+  type BuiltinOAuthProviderNames,
+  type TokenData,
+  type emailPasswordProviderName,
+} from "@edgedb/auth-core";
+import {
+  ClientAuth,
+  getConfig,
+  type AuthConfig,
+  type AuthOptions,
+} from "./client.js";
+
+export type { TokenData, AuthOptions, Client };
+
+export type BuiltinProviderNames =
+  | BuiltinOAuthProviderNames
+  | typeof emailPasswordProviderName;
+
+type ParamsOrError<Result extends object, ErrorDetails extends object = {}> =
+  | ({ error: null } & { [Key in keyof ErrorDetails]?: undefined } & Result)
+  | ({ error: Error } & ErrorDetails & { [Key in keyof Result]?: undefined });
+
+export interface AuthRouteHandlers {
+  onOAuthCallback(
+    params: ParamsOrError<{
+      tokenData: TokenData;
+      provider: BuiltinOAuthProviderNames;
+      isSignUp: boolean;
+    }>
+  ): Promise<Response>;
+  onBuiltinUICallback(
+    params: ParamsOrError<
+      (
+        | {
+            tokenData: TokenData;
+            provider: BuiltinProviderNames;
+          }
+        | { tokenData: null; provider: null }
+      ) & { isSignUp: boolean }
+    >
+  ): Promise<Response>;
+  onEmailVerify(
+    params: ParamsOrError<
+      { tokenData: TokenData },
+      { verificationToken?: string }
+    >
+  ): Promise<Response>;
+  onSignout(): Promise<Response>;
+}
+
+export default function serverAuth(client: Client, options: AuthOptions) {
+  const core = Auth.create(client);
+  const config = getConfig(options);
+
+  return {
+    createServerRequestAuth: ({ event }: { event: RequestEvent }) =>
+      new ServerRequestAuth(client, core, event, options),
+    createAuthRouteHook:
+      (handlers: AuthRouteHandlers): Handle =>
+      ({ event, resolve }) => {
+        const pathname = new URL(event.request.url).pathname;
+
+        if (pathname.startsWith(`/${config.authRoutesPath}`)) {
+          return createAuthRouteHandlers(handlers, event, core, config);
+        }
+
+        return resolve(event);
+      },
+  };
+}
+
+export class ServerRequestAuth extends ClientAuth {
+  private readonly client: Client;
+  private readonly core: Promise<Auth>;
+  private readonly cookies: Cookies;
+
+  public get session() {
+    return new AuthSession(
+      this.client,
+      this.cookies.get(this.config.authCookieName)
+    );
+  }
```

Comment:
Now this class is initialised per request, this can just be a property created in the constructor.

Reply:
This does make it lazy though, right? If you don't call `auth.session` it wouldn't create the class?

Reply:
Yes, but then you create it everytime you access session though. Ok, maybe it's fine as it is.

Reply:
Oh yeah I guess we can set it as an optional private property and return it if it's defined.

Reply:
Or, if we want to slightly optimise it, we could memoise it. I think I did that for auth-nextjs.

Reply:
hm is this ok?

```
const sessionCache: Record<string, AuthSession> = {};

export class ServerRequestAuth extends ClientAuth {
  private readonly client: Client;
  private readonly core: Promise<Auth>;
  private readonly cookies: Cookies;

  public get session() {
    const authCookie = this.cookies.get(this.config.authCookieName);

    if (authCookie && sessionCache[authCookie]) {
      return sessionCache[authCookie];
    }

    const authSession = new AuthSession(this.client, authCookie);

    if (authCookie) {
      sessionCache[authCookie] = authSession;
    }

    return authSession;
  }
```